From d11b9be5bc3f6b2bf924c2bb92f76e2c0b97bd7e Mon Sep 17 00:00:00 2001
From: Sai Praneeth <sai.praneeth.prakhya@intel.com>
Date: Thu, 9 Mar 2017 16:41:01 -0800
Subject: [PATCH 6/9] efi: Allow efi_mem_desc_lookup find illegally accessed
 EFI regions

CONFIG_EFI_WARN_ON_ILLEGAL_ACCESSES detects illegal accesses to EFI
regions using the offending physical address of the access. We need the
memory descriptor of such region so that it can be mapped as a fixup to
a page fault. These illegal accesses can occur in
EFI_CONVENTIONAL_MEMORY, EFI_LOADER_CODE/DATA or
EFI_BOOT_SERVICES_CODE/DATA regions. Thus, allow efi_mem_desc_lookup to
return the descriptor of any of the above regions.

Signed-off-by: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
Signed-off-by: Sai Praneeth Prakhya <sai.praneeth.prakhya@intel.com>
Cc: Lee, Chun-Yi <jlee@suse.com>
Cc: Borislav Petkov <bp@alien8.de>
Cc: Fenghua Yu <fenghua.yu@intel.com>
Cc: Ravi Shankar <ravi.v.shankar@intel.com>
Cc: Matt Fleming <matt@codeblueprint.co.uk>
Cc: Ard Biesheuvel <ard.biesheuvel@linaro.org>
---
 drivers/firmware/efi/efi.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/firmware/efi/efi.c b/drivers/firmware/efi/efi.c
index 76fe56d230bd..49c8df731fed 100644
--- a/drivers/firmware/efi/efi.c
+++ b/drivers/firmware/efi/efi.c
@@ -378,6 +378,12 @@ int __init_fixup efi_mem_desc_lookup(u64 phys_addr, efi_memory_desc_t *out_md)
 
 		if (!(md->attribute & EFI_MEMORY_RUNTIME) &&
 		    md->type != EFI_BOOT_SERVICES_DATA &&
+#if defined(CONFIG_EFI_WARN_ON_ILLEGAL_ACCESSES)
+		    md->type != EFI_BOOT_SERVICES_CODE &&
+		    md->type != EFI_CONVENTIONAL_MEMORY &&
+		    md->type != EFI_LOADER_CODE &&
+		    md->type != EFI_LOADER_DATA &&
+#endif
 		    md->type != EFI_RUNTIME_SERVICES_DATA) {
 			continue;
 		}
-- 
2.1.4

